desilike/theories/galaxy_clustering/power_template.py

```diff
@@ -1140,4 +1141,75 @@ def calculate(self, df=1., m=0.6, n=0.9, qto=1., dpto=1.):
     def get(self):
         self.DV_times_kTO = self.apeffect.qiso * self.DV_times_kTO_fid
         self.DH_over_DM = self.apeffect.qap * self.DH_over_DM_fid
-        return self
\ No newline at end of file
+        return self
+
+
+class WiggleRescalePowerSpectrumTemplate(BasePowerSpectrumTemplate):
+    """
+    Rescale the wiggly part of the power spectrum in order to marginalize over the sound horizon scale.
+
+    Parameters
+    ----------
+    k : array, default=None
+        Theory wavenumbers where to evaluate linear power spectrum.
+
+    z : float, default=1.
+        Effective redshift.
+
+    with_now : str, default='wallish2018'
+        Compute smoothed, BAO-filtered, linear power spectrum with this engine (e.g. 'wallish2018', 'peakaverage').
+
+    fiducial : str, tuple, dict, cosmoprimo.Cosmology, default='DESI'
+        Specifications for fiducial cosmology, used to compute the linear power spectrum. Either:
+
+        - str: name of fiducial cosmology in :class:`cosmoprimo.fiucial`
+        - tuple: (name of fiducial cosmology, dictionary of parameters to update)
+        - dict: dictionary of parameters
+        - :class:`cosmoprimo.Cosmology`: Cosmology instance
+
+    cosmo : BasePrimordialCosmology, default=None
+        Cosmology calculator. Defaults to ``Cosmoprimo(fiducial=fiducial)``.
+
+    Reference
+    ----------
+    - https://arxiv.org/abs/2112.10749
+    """
+    config_fn = 'power_template.yaml'
+    
+    def initialize(self, *args, k=None, z=1., with_now='wallish2018', fiducial='DESI', cosmo=None, **kwargs):
+        self.cosmo = cosmo
+        self.fiducial = fiducial
+        self.z = z
+        self.with_now = with_now
+        if k is None: k = np.logspace(-3., 1., 400)
+        self.k = np.array(k, dtype='f8')
+
+        # Make sure to use a larger k-grid than requested for the input power spectrum,
+        # so that after rescaling, the wiggles fully cover the desired k-range.
+        min_qbao = 0.49
+        max_qbao = 1.51
+        self.kin = np.logspace(np.log10(k[0]*min_qbao), np.log10(k[-1]*max_qbao), 400)
+        
+        super(WiggleRescalePowerSpectrumTemplate, self).initialize(*args, k=self.kin, z=self.z, with_now=self.with_now, fiducial=self.fiducial, apmode='distances', **kwargs)
+
+    def calculate(self, qbao=1.):
+        super(WiggleRescalePowerSpectrumTemplate, self).calculate()
+        
+        # Separate wiggles
+        self.pkw_dd = self.pk_dd - self.pknow_dd
+        
+        # Rescale k-grid for wiggles
+        kw_rescale = self.kin * qbao
+        
+        # Interpolate wiggle power spectrum to the original k-grid
+        pkw_dd_interpolator = interpolate.interp1d(kw_rescale, self.pkw_dd, kind='cubic')
```

Comment:
I think `self.k` and `self.kin` will be the same here. Best is probably to avoid defining kin. You can e.g. do:
```
k = self.pk_dd_interpolator_fid.k  # this is independent and much wider than self.k, typically
k = k[(k > k[0] * 2.) & (k < k[-1] / 2.)]  # to avoid hitting boundaries with qbao
wiggles = self.pk_dd_interpolator(k / qbao) - self.pknow_dd_interpolator(k / qbao)
# creating a new interpolator in case we need it (only used by BAO models)
self.pk_dd_interpolator = PowerSpectrumInterpolator1D(k, self.pk_dd_interpolator(k) + wiggles)
self.pk_dd = self.pk_dd_interpolator(self.k)
```

as in
https://github.com/cosmodesi/desilike/blob/6de28dfcf6218db80cf7a79d09c37ef3806d646d/desilike/theories/galaxy_clustering/power_template.py#L996